pynicotine/natpmp.py

```diff
@@ -0,0 +1,643 @@
+# COPYRIGHT (C) 2020-2022 Nicotine+ Contributors
+#
+# GNU GENERAL PUBLIC LICENSE
+#    Version 3, 29 June 2007
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from threading import Thread
+
+from pynicotine.config import config
+from pynicotine.events import events
+from pynicotine.logfacility import log
+
+RENEWAL_INTERVAL = 14400  # 4 hours
+
+import os
+import re
+import struct
+import socket
+import select
+import platform
+
+# Modified version of NAT-PMP client library from https://github.com/yimingliu/py-natpmp
+# 
+# Provides functions to interact with NAT-PMP gateways implementing version 0
+# of the NAT-PMP draft specification.
+# 
+# This version does not completely implement the draft standard.
+# * It does not provide functionality to listen for address change packets.
+# * It does not have a proper request queuing system, meaning that
+# multiple requests may be issued in parallel, against spec recommendations.
+# 
+# For more information on NAT-PMP, see the NAT-PMP draft specification:
+# http://files.dns-sd.org/draft-cheshire-nat-pmp.txt
+
+# Copyright (c) 2008-2016, Yiming Liu, All rights reserved.
```

Comment:
~`2016`~ -> `2023`

Please add to attributions sections in:

+ https://github.com/nicotine-plus/nicotine-plus/blob/master/AUTHORS.md#third-party-attributions

+ https://github.com/nicotine-plus/nicotine-plus/blob/07d5ca8ce878d508484b2942ce37ee37652d2e34/pynicotine/gtkgui/dialogs/about.py#L308

Also add yourself in somewhere if you want.